spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java

```diff
@@ -0,0 +1,272 @@
+    File location = temp.newFolder();
+    String dest = uniqueName("iceberg_snapshot_partitioned");
+    String source = uniqueName("test_snapshot_partitioned_table");
+    testCreate(source,
+        dest,
+        "CREATE TABLE %s using parquet PARTITIONED BY (id) location '%s' AS SELECT * FROM %s",
+        () -> Actions.snapshot(source, dest, location.toString()),
+        3);
+    testIsolatedSnapshot(source, dest);
+  }
+
+  @Test
+  public void testSnapshotUnpartitioned() throws Exception {
+    File location = temp.newFolder();
+    String dest = uniqueName("iceberg_snapshot_unpartitioned");
+    String source = uniqueName("test_snapshot_unpartitioned_table");
+    testCreate(source,
+        dest,
+        "CREATE TABLE %s using parquet location '%s' AS SELECT * FROM %s",
+        () -> Actions.snapshot(source, dest, location.toString()),
+        2);
+    testIsolatedSnapshot(source, dest);
+  }
+
+  @Test
+  public void testSnapshotHiveTable() throws Exception {
+    File location = temp.newFolder();
+    String dest = uniqueName("iceberg_snapshot_hive_table");
+    String source = uniqueName("snapshot_hive_table");
+    testCreate(source,
+        dest,
+        String.format("CREATE EXTERNAL TABLE %s (id Int, data String) STORED AS parquet LOCATION '%s'", source,
+            tableLocation),
+        () -> Actions.snapshot(source, dest, location.toString()),
+        3);
+    testIsolatedSnapshot(source, dest);
+  }
+
+  @Test
+  public void testProperties() throws Exception {
+    String dest = uniqueName("iceberg_properties");
+    String source = uniqueName("test_properties_table");
+    Map<String, String> props = Maps.newHashMap();
+    props.put("city", "New Orleans");
+    props.put("note", "Jazz");
+    testCreate(source,
+        dest,
+        "CREATE TABLE %s using parquet location '%s' AS SELECT * FROM %s",
+        () -> Actions.migrate(source).as(dest)
+        .withAdditionalProperty("dogs", "sundance")
+        .withAdditionalProperties(props),
+        2);
+
+    SparkTable table = loadTable(dest);
+
+
+    Map<String, String> expectedProps = Maps.newHashMap();
+    expectedProps.putAll(props);
+    expectedProps.put("dogs", "sundance");
+
+    for (Map.Entry<String, String> entry : expectedProps.entrySet()) {
+      Assert.assertTrue(
+          "Created table missing property " + entry.getKey(),
+          table.properties().containsKey(entry.getKey()));
+      Assert.assertEquals("Property value is not the expected value",
+          entry.getValue(), table.properties().get(entry.getKey()));
+    }
+  }
+
+  private SparkTable loadTable(String name) throws NoSuchTableException {
+    return (SparkTable) catalog.loadTable(Spark3Util.catalogAndIdentifier(spark, name).identifier());
+  }
+
+  // Creates a table, runs the migration command and checks the results.
+  private void testCreate(String source, String dest, String sqlCreate, Supplier<CreateAction> action,
+      long expectedMigratedFiles) throws
```

Comment:
I like the code reuse that these tests have, but I don't think that the pattern of a single method like this is very readable. For example, it isn't clear what `expectedMigratedFiles` is until you read this, but it differs between test cases.

I think a better pattern for reuse is to use separate methods that are well named. These tests have a great example of what I'm talking about with `testIsolatedSnapshot`. You could add a boolean to `testCreate` for whether this method should call `testIsolatedSnapshot`, but it is more readable to simply put the call in the test case directly.

I think a similar option, `assertMigratedFileCount`, would be an improvement over passing an unlabelled number. Similarly, building a method to create and initialize the source table with a name rather than passing a create statement would help readability in the tests.

Reply:
Sgtm